PES-1394 Page Add vendor

Mrakor opened the pull request.

Mrakor committed:
PES-1394 Add vendor template mock

Mrakor committed:
PES-1394 Template mock - add/remove weight rule row

improve design trifles
add some translations
temporary wip for rebase

Mrakor committed:
PES-1394 Save addVendor form data

Mrakor committed:
PES-1394 add new Vendor

Mrakor committed:
PES-1394 minor fixes

packeta-user commented on upload/admin/controller/extension/shipping/zasilkovna.php:
```diff
@@ -1365,4 +1367,136 @@ private function removeInvalidKeyFromPostData()
+	 */
+	public function add_vendor()
+	{
+		$data = $this->initPageData(self::ACTION_ADD_VENDOR, 'Nový vendor');
```
Nemá být text "Nový vendor" v překladech ? Použij ale "Nový dopravce" .  Vendor je zatím pouze naše interní označení.

packeta-user commented on upload/admin/controller/extension/shipping/zasilkovna.php:
```diff
@@ -1365,4 +1367,136 @@ private function removeInvalidKeyFromPostData()
+
+		//mock data
+		$vendorId = 106;
+		$data['weight_rules'] = $this->getVendorWeightRules($vendorId);
```
Proč načítáš weight_rules ? pro založení nového vendora to nepotřebuješ.  Jak jsme se dohodli, v rámci tohoto PR neřešíme editaci, jen založení nového vendora.

packeta-user commented on upload/admin/controller/extension/shipping/zasilkovna.php:
```diff
@@ -1365,4 +1367,136 @@ private function removeInvalidKeyFromPostData()
+	 * @return array
+	 */
+	private function getVendorsByCountry($countryCode)
+	{   //mock data
```
Sem dej TODO, že je potřeba dodělat načítání dopravců z DB, jak jsme se dohodli.

packeta-user commented on upload/admin/controller/extension/shipping/zasilkovna.php:
```diff
@@ -1365,4 +1367,136 @@ private function removeInvalidKeyFromPostData()
+	 *
+	 * @return array
+	 */
+	private function getVendorWeightRules($vendorId) {
```
Pro založení vendora tuto metodu nepotřebuješ.  Zatím ji tedy odstraň.

packeta-user replied:
tento komentář jsi uzavřel bez vyřešení.

packeta-user commented on upload/system/library/Packetery/Vendor/VendorRepository.php:
```diff
@@ -6,6 +6,18 @@
 
 class VendorRepository {
 
+	const INTERNAL_VENDORS = [
```
Chybí překlady.  To pole bych sestavil  takto 'id'=> 'zpoint', 'name'=> 'Výdejní místa',  countries=> ['cz', 'sk', 'ro', 'hu].  Bude to přehlednější a nemusíš několikrát opakovat názvy a idčka

packeta-user replied:
místo internal_vendors bych zvolil spíš název "packeta_vendors".

packeta-user commented on upload/system/library/Packetery/Vendor/VendorRepository.php:
```diff
@@ -37,30 +49,16 @@ public function getVendorsByCountry($country, $onlyEnabled = false) {
 				`zv`.`group`,
 				COALESCE(`zv`.`country`, `zc`.`country`) AS `country`,
-				IF(ISNULL(`zv`.`carrier_id`), 1, 0) AS 'has_pickup_points'
+				IF(ISNULL(`zv`.`carrier_id`), 1, `zc`.`is_pickup_points`) AS 'has_pickup_points'
```
Nejsem si 100% jistý, ale nebyla tato změna v jiném PR ?

packeta-user commented on upload/system/library/Packetery/Vendor/VendorRepository.php:
```diff
@@ -71,4 +69,163 @@ public function getVendorsByCountry($country, $onlyEnabled = false) {
+	 *
+	 * @return array
+	 */
+	private function updateVendor(array $vendor) {
```
jak jsme se dohodli,  aktualizaci vendora v rámci ticketu neřešíme.  toto tedy nyní nepotřebujeme.

packeta-user commented on upload/system/library/Packetery/Vendor/VendorRepository.php:
```diff
@@ -71,4 +69,163 @@ public function getVendorsByCountry($country, $onlyEnabled = false) {
+	 *
+	 * @return array
+	 */
+	public function deleteVendorPricesByIds(array $vendorPriceIds) {
```
odstranění vendora také teď neřešíme.

Mrakor committed:
PES-1394 Cr fixes

remove logic that doesn't belong to add carrier action
add translations
refactor INTERNAl_VENDORS and rename to PACKETA_VENDORS

packeta-user commented on upload/system/library/Packetery/Vendor/VendorRepository.php:
```diff
@@ -18,7 +24,7 @@ public function __construct(DB $db) {
 	/**
 	 * @param string $country
-	 * @param bool $onlyEnabled
+	 * @param bool   $onlyEnabled
```
není tam nějak moc mezer ?

Mrakor committed:
PES-1394 CR fix - remove unused method, CS fix

Mrakor committed:
PES-1394 For testing with debug info

Mrakor committed:
PES-1394 Remove debug info

Mrakor committed:
PES-1394 minor fixes

packeta-user commented on upload/admin/language/cs-cz/extension/shipping/zasilkovna.php:
```diff
@@ -192,3 +192,25 @@
+$_['vendor_add_carriers_optgroup'] = 'Dopravci';
+$_['vendor_add_vendor_label'] = 'Dopravce';
+$_['vendor_add_entry_cart_name'] = 'Název dopravce v košíku';
+$_['vendor_add_help_cart_name'] = 'Tooltip pro nazev dopravce v kosiku'; //TODO
```
TODO musí vždy nějaký popis, aby bylo jasné, co je potřeba dodělat.  Pokud nevíš jakou textaci použít , je nutné to konzultovat během vývoje a ne to předat na CR a čekat, až si toho někdo všimne a řekne ti, co tam má být.  Textace v tomto případě bude:  Název dopravce, který bude uveden v pokladně.

packeta-user commented on upload/admin/controller/extension/shipping/zasilkovna.php:
```diff
@@ -1370,4 +1363,76 @@ private function removeInvalidKeyFromPostData()
 	}
 
+	/**
+	 * @throws ReflectionException
```
chybí @return void

packeta-user commented on upload/admin/view/template/extension/shipping/zasilkovna_add_vendor.twig:
```diff
@@ -0,0 +1,195 @@
+</div>
+<script type="text/javascript"><!--
+    var weight_rule_row = {{ weight_rules|length }};
+    function add() {
```
Co když bude funkce se stejným názvem deklarována i jiným pluginem třetí strany a bude také načítána v administraci.  Neudělá to problém ?

packeta-user commented on upload/admin/view/template/extension/shipping/zasilkovna_add_vendor.twig:
```diff
@@ -0,0 +1,195 @@
+    </div>
+</div>
+<script type="text/javascript"><!--
+    var weight_rule_row = {{ weight_rules|length }};
```
Pokud je proměnná deklarovaná v obecném prostoru,  neměli bychom ji prefixovat, aby nedošlo k případné kolizi názvů s jiným modulem ?

packeta-user commented on upload/system/library/Packetery/Carrier/CarrierRepository.php:
```diff
@@ -186,5 +186,13 @@ public function getZpointCountryCodes()
 		];
 	}
+
+	public function getCarriersByCountry($countryCode)
```
Chybí DOC !!!!

packeta-user commented on upload/system/library/Packetery/Vendor/AddEditPage.php:
```diff
@@ -0,0 +1,335 @@
+		$this->session->data['vendor_add_form_data'] = $formData;
+	}
+
+	private function getAddVendorFormDataFromSession() {
```
Chybí DOC !!!!

packeta-user commented on upload/system/library/Packetery/Vendor/VendorRepository.php:
```diff
@@ -57,4 +63,105 @@ public function getVendorsByCountry($country, $onlyEnabled = false) {
+			$this->db->escape($vendor['country']),
+			$this->db->escape($vendor['group']),
+			$vendor['free_shipping_limit'] ?: 'NULL',
+			($vendor['max_weight'] === null) ? 'NULL' : $vendor['max_weight'],
```
Klíč  "max_weight" v poli $vendor neexistuje.  Zde plugin vypíše chybu: Notice: Undefined index: max_weight in ....

packeta-user commented on upload/system/library/Packetery/Vendor/VendorRepository.php:
```diff
@@ -57,4 +63,105 @@ public function getVendorsByCountry($country, $onlyEnabled = false) {
+
+		$sql = sprintf(
+			"INSERT INTO `%s`
+			(`carrier_id`, `carrier_name_cart`,`country`, `group`, `free_shipping_limit`, `max_weight`, `is_enabled`)
```
Sloupec max_weight není součástí tabulky zasilkovna_vendor. Toto způsobí, že při pokusu o uložení modul spadne na chybě "Fatal error: Uncaught Exception: Error: Unknown column 'max_weight' ".

packeta-user commented on upload/system/library/Packetery/Vendor/AddEditPage.php:
```diff
@@ -0,0 +1,335 @@
+		$requiredFields = ['vendor'];
+
+		if ($formData['action'] === ControllerExtensionShippingZasilkovna ::ACTION_ADD_VENDOR) {
+			foreach ($requiredFields as $requiredField) {
```
Nevidím smysl iterovat pole o jednom prvku.  Foreach je v tomto případně zbytečný.  Stačí jen podmínka 
if (empty($formData['vendor'])) { ....

packeta-user commented on upload/admin/controller/extension/shipping/zasilkovna.php:
```diff
@@ -1370,4 +1363,76 @@ private function removeInvalidKeyFromPostData()
+
+		$postedData = $this->request->post;
+
+		if (!empty($postedData) && $postedData['action'] === self::ACTION_ADD_VENDOR) {
```
Pro zjistění zda byl formulář odeslaný a má se zpracovat používáme v celém pluginu podmínku:
if ($this->request->server['REQUEST_METHOD'] === 'POST') {
Nemá cenu vymýšlet něco jiného na jednom konkrétním místě.   Udělal bych to zde stejně.

packeta-user commented on upload/system/library/Packetery/Vendor/AddEditPage.php:
```diff
@@ -0,0 +1,335 @@
+		$weightRulesErrors = [];
+		$requiredFields = ['vendor'];
+
+		if ($formData['action'] === ControllerExtensionShippingZasilkovna ::ACTION_ADD_VENDOR) {
```
Tady bych nejdřív vyřešil if ($formData['action'] !== ControllerExtensionShippingZasilkovna ::ACTION_ADD_VENDOR) ...  
Zbytek logiky pak nebudeš mít zanořený v if {} a kód bude přehlednější.

packeta-user replied:
Zde: "ControllerExtensionShippingZasilkovna ::ACTION_ADD_VENDOR" máš chybně napsanou konstantu - před :: máš tam mezeru! PhPStorm takové chyby červeně podtrhává a říká, že konstanta neexistuje (Constant 'ACTION_ADD_VENDOR' not found in \ControllerExtensionShippingZasilkovna).

packeta-user commented on upload/admin/language/cs-cz/extension/shipping/zasilkovna.php:
```diff
@@ -192,3 +192,25 @@
+$_['vendor_add_packeta_optgroup'] = 'Zásilkovna';
+$_['vendor_add_carriers_optgroup'] = 'Dopravci';
+$_['vendor_add_vendor_label'] = 'Dopravce';
+$_['vendor_add_entry_cart_name'] = 'Název dopravce v košíku';
```
Název dopravce v pokladně.

packeta-user commented on upload/system/library/Packetery/Vendor/AddEditPage.php:
```diff
@@ -0,0 +1,335 @@
+				continue;
+			}
+			$internalVendors[] = [
+				'vendor_id' => ($vendor['id'] === 'zpoint') ? '' : $vendor['id'],
```
Zde připravuješ pole ze kterého pak v šabloně sestavuješ select pro výběr dopravců.  Nemůžeš mít vendor_id prázdný string.  Prázdný string už používáš v selectu v promptu.  Pokud pak zvolíš dopravce "Packeta výdejní místo" neprojde dopravce validací a dopravce nejde uložit, protože ve validaci kontroluješ :
if (empty($formData[$requiredField]))

packeta-user commented on upload/admin/view/template/extension/shipping/zasilkovna_add_vendor.twig:
```diff
@@ -0,0 +1,195 @@
+                                        {% if vendors %}
+                                            <optgroup label="{{ carriers_optgroup }}">
+                                            {% for vendor in vendors %}
+                                                {% if vendor.vendor_id == vendor %}
```
Tato podmínka nebude nikdy splněna. Nemůžeš porovnávat vendor (array) s vendor.vendor_id (prvek pole). výsledek bude vždy false.

packeta-user replied:
v poli vendor nemáš prvek "vendor_id", ale "id".  Ve value jednotlivých option tak nebudeš mít žádnou hodnotu. Formulář po odeslání nemůže projít validací a bude hlásit, že dopravce nebyl vybrán.

Mrakor committed:
PES-1394 CR fixes

correct translations
add missing annotations
fix template - packeta specific JS
fix functionality

Mrakor committed:
PES-1394 CR fix - add missing annotation, better logic

packeta-user commented on upload/admin/view/template/extension/shipping/zasilkovna_add_vendor.twig:
```diff
@@ -0,0 +1,195 @@
+</div>
+<script type="text/javascript"><!--
+    var packeta_weight_rule_row_count = {{ weight_rules|length }};
+    function addWeightRuleRow() {
```
Co když bude funkce se stejným názvem deklarována i jiným pluginem třetí strany a bude také načítána v globálním prostoru, jako ji máš ty ? Neudělá to problém ?

packeta-user commented on upload/admin/controller/extension/shipping/zasilkovna.php:
```diff
@@ -1370,4 +1363,77 @@ private function removeInvalidKeyFromPostData()
+
+		$postedData = $this->request->post;
+
+		if (($this->request->server['REQUEST_METHOD'] === 'POST') && ($postedData['action'] === self::ACTION_ADD_VENDOR)) {
```
toto je v podmínce zbytečné: && ($postedData['action'] === self::ACTION_ADD_VENDOR)

packeta-user commented on upload/system/library/Packetery/Carrier/CarrierRepository.php:
```diff
@@ -186,5 +186,18 @@ public function getZpointCountryCodes()
+	 *
+	 * @return mixed
+	 */
+	public function getCarriersByCountry($countryCode)
```
U této metody uvádíš, že může vrátit mixed.  Výsledek této metody ale předáváš jako první parametr array_filter(), což musí být array.  Buď máš špatně návratový typ, nebo nemáš ošetřeno, abys nepředal array_filter chybný parametr.

packeta-user commented on upload/system/library/Packetery/Vendor/AddEditPage.php:
```diff
@@ -0,0 +1,340 @@
+	/**
+	 * @return array
+	 */
+	public function getTranslations() {
```
Proč potřebuješ sestavit pole s překlady a předávat je šabloně ?  Twig si natahuje překlady automaticky.  V šabloně by ti mělo stačit zavolat např.  {{vendor_add_vendor_label}} a dostaneš překlad ze slovníku.

packeta-user commented on upload/system/library/Packetery/Vendor/AddEditPage.php:
```diff
@@ -0,0 +1,340 @@
+		$weightRulesErrors = [];
+
+		if ($formData['action'] === ControllerExtensionShippingZasilkovna::ACTION_ADD_VENDOR) {
+			if (empty($formData['vendor'])) {
```
Pokud ve formuláři nevyberu žádného dopravce, tak $formData['vendor'] bude "-1".  Takže to projde touto podmínkou a vytvoří se vendor , kde carrier_id je -1.

packeta-user commented on upload/system/library/Packetery/Vendor/AddEditPage.php:
```diff
@@ -0,0 +1,340 @@
+		}
+		$prefilledData = [];
+		foreach ($formData as $key => $value) {
+			if (in_array($key, ['vendor', 'country', 'vendor_group', 'cart_name', 'free_shipping_limit', 'is_enabled', 'weight_rules'])) {
```
Toto není dobrý způsob řešení.  Musíš dávat pozor, aby jsi v tomto poli neměl stejný klíč jako v poli $data v metodě add_vendor(), protože tam děláš array_merge($data, $prefilledData).   Druhým důvodem je to, že když v budoucnu rozšíříš formulář o další pole, tak si budeš muset pamatovat,  že musíš jít do této metody a zde upravit obsah pole.  To je velice náchylné na chyby.

packeta-user commented on upload/admin/view/template/extension/shipping/zasilkovna_add_vendor.twig:
```diff
@@ -0,0 +1,195 @@
+                                <label class="col-sm-2 control-label"  for="input-vendor">{{ vendor_label }}</label>
+                                <div class="col-sm-10">
+                                    <select name="vendor" id="input-vendor" class="form-control">
+                                        <option value="-1">{{ text_select_vendor }}</option>
```
Pokud není v selectu vybraná hodnota, nastav value na prázdný string a ošetři v PHP.  Použít hodnot "-1" je velice nešťastné a vede to k chybám.

Mrakor committed:
PES-1394 CR fixes

remove redundant translation data
add correct translations keys to template
isolate javascript

Mrakor committed:
PES-1394 Fix - display warning if no country is part of URL

Mrakor committed:
PES-1394 Fixed breadcrumbs for carrier settings and subpages

Mrakor committed:
PES-1394 Fix saving empty weight rule

Mrakor committed:
PES-1394 Fix carrier_settings_country if no country code in URL

Mrakor committed:
PES-1394 CR refactor

remove session from AddEditPage
isolate form data in template data
remove inline JS event handlers
remove sorting of weight rules
PES-1394 fixup


PES-1394 CS fixup

packeta-user commented on upload/system/library/Packetery/Vendor/AddEditPage.php:
```diff
@@ -0,0 +1,227 @@
+	 *
+	 * @return array
+	 */
+	public function validateForm(array $postedData) {
```
přejmenuj na validate()

packeta-user commented on upload/system/library/Packetery/Vendor/AddEditPage.php:
```diff
@@ -0,0 +1,227 @@
+		$validationErrors = $this->getAddVendorFormValidationErrors($postedData);
+		if (!empty($validationErrors)) {
+			foreach ($validationErrors as $error) {
+				$errors[] = $this->language->get($error);
```
Proč překládáš už jednou přeložený řetězec ?

Mrakor committed:
PES-1394 Error warnings at appropriate form fields


PES-1394 fixup

packeta-user commented on upload/system/library/Packetery/Vendor/AddEditPage.php:
```diff
@@ -0,0 +1,189 @@
+		}
+
+		$vendorPrices = [];
+		if ($formData['weight_rules'] && is_array($formData['weight_rules'])) {
```
V jakém případě bude tato podmínka vyhodnocena jako false ? Toto by mělo být ošetřeno už při validaci.  Pokud by byla podmínka vyhodnocena jako false,  bude vendor v DB nekompletní (bude existovat záznam v tabulce vendor, ale nebude existovat záznam v tabulce vendor_price).

packeta-user commented on upload/system/library/Packetery/Vendor/AddEditPage.php:
```diff
@@ -0,0 +1,189 @@
+					'price'      => (float)$weightRule['price'],
+				];
+			}
+			if ($vendorPrices) {
```
Ta samá otázka jako v předchozím komentáři .  O validaci vendora by se měla starat validace.  Při ukládání je další validace zbytečná.

packeta-user commented on upload/system/library/Packetery/Vendor/AddEditPage.php:
```diff
@@ -0,0 +1,189 @@
+	 *
+	 * @return array
+	 */
+	public function getCarriersByCountry($countryCode) {
```
Co dělá tato metoda ? Podle názvu bych čekal, že jako výsledek dostanu všechny dopravce konkrétní země.

packeta-user commented on upload/system/library/Packetery/Vendor/VendorRepository.php:
```diff
@@ -57,4 +63,104 @@ public function getVendorsByCountry($country, $onlyEnabled = false) {
+	 *
+	 * @return array
+	 */
+	public function getInternalVendorsByCountry($countryCode) {
```
Proč je tato metoda v repositáři ?

Mrakor replied:
za me sem patri.  repozitar chapu jako centralni misto pro praci s nejakym typem entity.

xpavp03 commented on upload/admin/controller/extension/shipping/zasilkovna.php:
```diff
@@ -1448,7 +1452,8 @@ public function add_vendor() {
 			if (!empty($data['form']['weight_rules'])) {
 				$data['form']['weight_rules'] = $vendorAddEditPage->removeEmptyWeightRules($data['form']['weight_rules']);
```
Proč tady voláš removeEmptyWeightRules() ještě jednou?

xpavp03 commented on upload/system/library/Packetery/Vendor/AddEditPage.php:
```diff
@@ -97,45 +95,26 @@ public function getAddVendorFormValidationErrors(array $formData) {
 
-			if (!isset($rule['price']) || !is_numeric($rule['price']) || $rule['price'] <= 0) {
-				$errors[] = 'vendor_add_error_rule_price_invalid';
+			if (!is_numeric($rule['price']) || $rule['price'] <= 0) {
```
nechceme povolit cenu 0?
váhu 0 ne, to chápu, ale třeba chce dopravu zadarmo pro váhu do nějakého limitu... Zeptej se Jardy, prosím.

packeta-user replied:
Eshopy chtějí dopravu zdarma od určité částky.  Dopravu zdarma podle hmotnosti jsem nikdy neviděl.  V původním nastavení modulu to také nejde nastavit a žádný eshop nám nikdy nepsal, že by něco takového chtěl.  Takže bych to nechal jak toje.

packeta-user commented on upload/system/library/Packetery/Vendor/AddEditPage.php:
```diff
@@ -0,0 +1,189 @@
+
+use Packetery\Carrier\CarrierRepository;
+
+class AddEditPage {
```
Přejmenoval bych to na "Page".

packeta-user commented on upload/system/library/Packetery/Vendor/VendorRepository.php:
```diff
@@ -57,4 +63,104 @@ public function getVendorsByCountry($country, $onlyEnabled = false) {
+			return $this->db->getLastId();
+		}
+
+		return null;
```
V jakém případě může nastat tato situace a metoda vrací null ?

packeta-user commented on upload/system/library/Packetery/Vendor/VendorRepository.php:
```diff
@@ -57,4 +63,104 @@ public function getVendorsByCountry($country, $onlyEnabled = false) {
+
+		$queryResult =$this->db->query($sql);
+		$groups = [];
+		if ($queryResult->num_rows === 0) {
```
Proč je zde tato podmínka ?

packeta-user commented on upload/system/library/Packetery/Vendor/VendorRepository.php:
```diff
@@ -57,4 +63,104 @@ public function getVendorsByCountry($country, $onlyEnabled = false) {
+			return $groups;
+		}
+
+		foreach($queryResult->rows as $row) {
```
Není to foreach zbytečné ? Nestačilo by return array_column($queryResult->rows, 'group')

packeta-user commented on upload/system/library/Packetery/Vendor/VendorRepository.php:
```diff
@@ -6,6 +6,12 @@
 class VendorRepository {
 
+	const PACKETA_VENDORS = [
+		['id' => 'zpoint', 'name' => 'vendor_add_zpoint', 'countries' => ['cz', 'sk', 'hu', 'ro']],
```
Za posledním prvkem pole píšeme vždy ","

Mrakor committed:
PES-1394 CR fixes

remove redundant code
rename method for clarity
cs fixes

Mrakor committed:
PES-1394 fix - reindex array with weight rules

Mrakor committed:
PES-1394 Fix notice when no weight rules added

Commit:
CR fixes and refactoring

Commit:
PES-1394 fix: name_cart -> cart_name in table zasilkovna_vendor
